File: apps/vulnfounder-cli/internal/checkpoint/checkpoint.go
// Package checkpoint provides auto-resume detection for LLM pipeline steps.
//
// When a step (enhance, analyze, verify) is interrupted, per-unit checkpoint
// files remain in {scanDir}/{step}_checkpoints/. On the next run the Go CLI
// detects these files and prompts the user to resume or start fresh.
//
// Checkpoint status (completed vs errored counts) is determined by calling
// the Python CLI's `checkpoint-status` command, which is the single source
// of truth for checkpoint semantics.
package checkpoint

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/fatih/color"
	"github.com/synbnb/vulnfounder/apps/vulnfounder-cli/internal/python"
)

const summaryFile = "_summary.json"

// fingerprintFile is the backend-identity sidecar (I2) the Python pipeline
// writes into each checkpoint dir. Like _summary.json it is NOT a per-unit
// result and must be excluded from the fallback count, or a resume would
// over-count "completed" units by one.
const fingerprintFile = "_fingerprint.json"

// isSidecar reports whether a checkpoint-dir entry is a bookkeeping sidecar
// (never a per-unit result) and so must not be counted as a completed unit.
func isSidecar(name string) bool {
	return name == summaryFile || name == fingerprintFile
}

// Summary represents the _summary.json written by Python pipeline steps.
type Summary struct {
	Step           string         `json:"step"`
	Phase          string         `json:"phase"`
	Timestamp      string         `json:"timestamp"`
	TotalUnits     int            `json:"total_units"`
	Completed      int            `json:"completed"`
	Errors         int            `json:"errors"`
	ErrorBreakdown map[string]int `json:"error_breakdown"`
}

// Info describes an existing checkpoint directory.
type Info struct {
	Dir     string   // full path to the checkpoint dir
	Count   int      // number of successfully completed units
	Errors  int      // number of errored units
	Summary *Summary // parsed _summary.json (may have counts overridden by Python)
}

// Phase returns the detected phase state as a human-readable string.
func (i *Info) Phase() string {
	if i.Summary == nil {
		return "legacy"
	}
	if i.Summary.Phase == "done" && i.Summary.Errors > 0 {
		return "done_with_errors"
	}
	return i.Summary.Phase // "in_progress" or "done"
}

// DetectViaPython checks for checkpoints by calling the Python CLI's
// checkpoint-status command for accurate completed/errored counts.
// Returns nil if no checkpoint is found or Python fails.
func DetectViaPython(pythonPath, scanDir, stepName string) *Info {
	dir := filepath.Join(scanDir, stepName+"_checkpoints")

	// Quick filesystem check: skip Python call if dir doesn't exist
	if _, err := os.Stat(dir); err != nil {
		return nil
	}

	// Call Python for accurate counts
	result, err := python.Invoke(pythonPath, []string{"checkpoint-status", dir}, "", true, "")
	if err != nil || result.Envelope.Status != "success" {
		// Python failed — fall back to simple file count
		return DetectFallback(scanDir, stepName)
	}

	// Parse the response data
	dataBytes, err := json.Marshal(result.Envelope.Data)
	if err != nil {
		return DetectFallback(scanDir, stepName)
	}

	var status struct {
		Step           string         `json:"step"`
		Completed      int            `json:"completed"`
		Errors         int            `json:"errors"`
		TotalFiles     int            `json:"total_files"`
		TotalUnits     int            `json:"total_units"`
		Phase          string         `json:"phase"`
		ErrorBreakdown map[string]int `json:"error_breakdown"`
	}
	if json.Unmarshal(dataBytes, &status) != nil {
		return DetectFallback(scanDir, stepName)
	}

	if status.TotalFiles == 0 {
		return nil
	}

	// If the previous run finished cleanly (phase=done, no errors), there's
	// nothing to resume — treat it as if there are no checkpoints. The
	// checkpoint files are preserved for audit/retro but don't trigger a prompt.
	if status.Phase == "done" && status.Errors == 0 {
		return nil
	}

	info := &Info{
		Dir:    dir,
		Count:  status.Completed,
		Errors: status.Errors,
		Summary: &Summary{
			Step:           status.Step,
			Phase:          status.Phase,
			TotalUnits:     status.TotalUnits,
			Completed:      status.Completed,
			Errors:         status.Errors,
			ErrorBreakdown: status.ErrorBreakdown,
		},
	}

	return info
}

// DetectFallback checks for checkpoints using only filesystem scanning.
// Used when Python is not available. Counts .json files without classifying
// completed vs errored — all files are counted as completed.
func DetectFallback(scanDir, stepName string) *Info {
	dir := filepath.Join(scanDir, stepName+"_checkpoints")

	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}

	count := 0
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".json") && !isSidecar(e.Name()) {
			count++
		}
	}
	if count == 0 {
		return nil
	}

	info := &Info{Dir: dir, Count: count}

	// Try to read _summary.json for phase state and total_units
	summaryPath := filepath.Join(dir, summaryFile)
	data, err := os.ReadFile(summaryPath)
	if err == nil {
		var s Summary
		if json.Unmarshal(data, &s) == nil {
			info.Summary = &s
		}
	}

	return info
}

// PromptResume asks the user whether to resume an interrupted run or start
// fresh. Returns true for resume, false for fresh start.
//
// In non-interactive mode (stdin is not a terminal, or quiet mode), defaults
// to resume.
func PromptResume(info *Info, stepName string, quiet bool) bool {
	if quiet || !isTerminal() {
		// Non-interactive: default to resume
		dim := color.New(color.Faint)
		dim.Fprintf(os.Stderr, "[%s] Auto-resuming from %d checkpointed units (non-interactive)\n",
			stepName, info.Count)
		return true
	}

	yellow := color.New(color.FgYellow, color.Bold)
	red := color.New(color.FgRed, color.Bold)
	bold := color.New(color.Bold)

	fmt.Fprintln(os.Stderr)

	switch info.Phase() {
	case "in_progress":
		// Interrupted run — show progress out of total
		yellow.Fprintf(os.Stderr, "Previous %s run interrupted", stepName)
		s := info.Summary
		if info.Errors > 0 {
			fmt.Fprintf(os.Stderr, " (%d/%d completed, %d errors)\n",
				info.Count, s.TotalUnits, info.Errors)
		} else {
			fmt.Fprintf(os.Stderr, " (%d/%d completed)\n",
				info.Count, s.TotalUnits)
		}

	case "done_with_errors":
		// Ran to completion but had errors — different message
		red.Fprintf(os.Stderr, "Previous %s run completed with errors", stepName)
		s := info.Summary
		fmt.Fprintf(os.Stderr, " (%d/%d completed, %d errors)\n",
			info.Count, s.TotalUnits, info.Errors)

	case "done":
		// Clean completion — shouldn't normally get here (checkpoints cleaned up)
		yellow.Fprintf(os.Stderr, "Previous %s run completed", stepName)
		s := info.Summary
		fmt.Fprintf(os.Stderr, " (%d/%d completed)\n", info.Count, s.TotalUnits)

	default:
		// Legacy format (no _summary.json) or fallback
		yellow.Fprintf(os.Stderr, "Previous %s run found", stepName)
		if info.Errors > 0 {
			fmt.Fprintf(os.Stderr, " (%d completed, %d errors)\n", info.Count, info.Errors)
		} else {
			fmt.Fprintf(os.Stderr, " (~%d units)\n", info.Count)
		}
	}

	fmt.Fprintf(os.Stderr, "  Checkpoint: %s\n", info.Dir)
	fmt.Fprintln(os.Stderr)
	bold.Fprint(os.Stderr, "Resume where you left off? ")
	fmt.Fprint(os.Stderr, "[Y/n] (n = discard progress, start fresh) ")

	reader := bufio.NewReader(os.Stdin)
	answer, _ := reader.ReadString('\n')
	answer = strings.TrimSpace(strings.ToLower(answer))

	if answer == "" || answer == "y" || answer == "yes" {
		return true
	}
	return false
}

// Clean removes a checkpoint directory.
func Clean(dir string) error {
	return os.RemoveAll(dir)
}

// isTerminal checks if stdin is a terminal (not a pipe).
func isTerminal() bool {
	fi, err := os.Stdin.Stat()
	if err != nil {
		return false
	}
	return fi.Mode()&os.ModeCharDevice != 0
}

File: libs/vulnfounder-core/core/checkpoint.py
"""
Shared checkpoint utilities for resumable pipeline steps.

Each LLM-heavy step (enhance, analyze, verify) can save per-unit checkpoint
files so interrupted runs resume where they left off. The checkpoint dir
lives next to the output file:

    {scan_dir}/enhance_checkpoints/
    {scan_dir}/analyze_checkpoints/
    {scan_dir}/verify_checkpoints/

On success (all units done), the checkpoint dir is cleaned up automatically.

Usage:

    cp = StepCheckpoint("enhance", output_dir="/path/to/scan/dir")
    completed = cp.load()                 # set of unit IDs already done
    ...process units...
    cp.save(unit_id, data_dict)           # save one unit
    cp.cleanup()                          # remove dir on success
"""

import json
import os
import shutil
import sys
from datetime import datetime, timezone

from utilities.safe_filename import safe_filename
from utilities.file_io import read_json, write_json
from core.backend_identity import FINGERPRINT_FILE
from pathlib import Path


SUMMARY_FILE = "_summary.json"

# Sidecar files that live inside a checkpoint dir but are NOT per-unit results.
# They must be excluded from every checkpoint counter (exists/count/status/load)
# and from the Go DetectFallback scan, or a resume would over-count "completed"
# units by the number of sidecars.
_RESERVED_FILES = frozenset({SUMMARY_FILE, FINGERPRINT_FILE})


class StepCheckpoint:
    """Manages per-unit checkpoint files for a pipeline step."""

    def __init__(self, step_name: str, output_dir: str):
        """
        Args:
            step_name: Pipeline step name (enhance, analyze, verify).
            output_dir: Directory where step outputs live (scan dir).
        """
        self.step_name = step_name
        self.dir = os.path.join(output_dir, f"{step_name}_checkpoints")

    @property
    def exists(self) -> bool:
        """True if a checkpoint directory exists with at least one unit file."""
        if not os.path.isdir(self.dir):
            return False
        return any(f.endswith(".json") and f not in _RESERVED_FILES
                   for f in os.listdir(self.dir))

    def count(self) -> int:
        """Number of per-unit checkpoint files (excludes _summary.json)."""
        if not os.path.isdir(self.dir):
            return 0
        return sum(1 for f in os.listdir(self.dir)
                   if f.endswith(".json") and f not in _RESERVED_FILES)

    def ensure_dir(self):
        """Create the checkpoint directory if it doesn't exist."""
        os.makedirs(self.dir, exist_ok=True)

    def load(self) -> dict[str, dict]:
        """Load all checkpointed units.

        Returns:
            Dict mapping unit_id -> checkpoint data dict.
        """
        results = {}
        if not os.path.isdir(self.dir):
            return results

        for filename in os.listdir(self.dir):
            if not filename.endswith(".json") or filename in _RESERVED_FILES:
                continue
            filepath = os.path.join(self.dir, filename)
            try:
                data = read_json(filepath)
                unit_id = data.get("id")
                if unit_id:
                    results[unit_id] = data
            except (json.JSONDecodeError, OSError):
                continue

        return results

    def load_ids(self, skip_errors: bool = True) -> set[str]:
        """Load just the set of completed unit IDs.

        Args:
            skip_errors: If True, don't count units that errored as completed.
                Supports all four phase formats: enhance, analyze, verify, dynamic-test.
        """
        ids = set()
        loaded = self.load()
        for unit_id, data in loaded.items():
            if skip_errors:
                # Enhance: agent_context.error
                agent_ctx = data.get("agent_context", {})
                if agent_ctx.get("error"):
                    continue
                # Analyze: result.verdict/finding
                result = data.get("result", {})
                if result.get("verdict") == "ERROR" or result.get("finding") == "error":
                    continue
                # Verify: verification empty or correct_finding == "error"
                if "verification" in data:
                    v = data.get("verification", {})
                    if not v or v.get("correct_finding") == "error":
                        continue
                # Dynamic-test: top-level status == "ERROR"
                if data.get("status") == "ERROR":
                    continue
            ids.add(unit_id)
        return ids

    def save(self, unit_id: str, data: dict):
        """Save a single unit's checkpoint.

        Args:
            unit_id: The unit identifier.
            data: Dict to persist (must include 'id' key).
        """
        self.ensure_dir()
        filename = self._safe_filename(unit_id) + ".json"
        filepath = os.path.join(self.dir, filename)
        data["id"] = unit_id  # ensure id is always present
        write_json(filepath, data)

    # ------------------------------------------------------------------
    # Backend-identity adopt gate (I2, minimal)
    # ------------------------------------------------------------------

    def sync_identity(self, fingerprint: dict, *, verbose: bool = True) -> dict:
        """Gate checkpoint adoption on backend identity.

        Call AFTER ``self.dir`` is finalized (respect any dir override, e.g.
        the verifier's ``verify_checkpoints``) and BEFORE any ``load()`` — it
        decides whether the prior checkpoints in this dir may be adopted by the
        current backend, and if not it archives them aside and starts fresh.

        Returns ``{"status", "archived_to", "warnings"}`` where status is one
        of:

          * ``new``    — no sidecar and no units: stamp and proceed.
          * ``legacy`` — units present but no sidecar (a pre-feature dir): we
            cannot verify identity, so ADOPT (never spuriously re-pay) and
            stamp for next time.
          * ``match``  — sidecar KEY digest equals the current one: adopt.
          * ``reset``  — sidecar KEY mismatch OR a corrupt/unreadable sidecar:
            archive the whole dir aside (preserve-not-destroy) and start fresh
            so the new backend re-pays.

        FAIL-CLOSED: a corrupt / unreadable sidecar is treated as ``reset``
        (archive + re-run), NEVER adopt — an unverifiable identity must not
        silently serve another backend's verdicts.
        """
        sidecar_path = os.path.join(self.dir, FINGERPRINT_FILE)
        has_units = self.exists
        warnings: list[str] = []

        sidecar_present = os.path.isfile(sidecar_path)
        existing = None
        corrupt = False
        if sidecar_present:
            try:
                existing = read_json(sidecar_path)
                if not isinstance(existing, dict):
                    corrupt = True
            except (json.JSONDecodeError, OSError, ValueError):
                corrupt = True

        # Corrupt / unreadable sidecar → FAIL CLOSED (archive + re-run).
        if corrupt:
            archived_to = self._archive_dir(fingerprint.get("key_digest", ""))
            os.makedirs(self.dir, exist_ok=True)
            self._write_fingerprint(fingerprint)
            msg = ("checkpoint identity sidecar was unreadable/corrupt; refusing "
                   "to adopt possibly-stale checkpoints (fail-closed). Archived "
                   f"to {archived_to}. Re-running this phase.")
            warnings.append(msg)
            if verbose:
                print(f"[{self.step_name}] {msg}", file=sys.stderr)
            return {"status": "reset", "archived_to": archived_to,
                    "warnings": warnings}

        # No sidecar recorded.
        if not sidecar_present:
            self._write_fingerprint(fingerprint)
            status = "legacy" if has_units else "new"
            if status == "legacy" and verbose:
                print(f"[{self.step_name}] Adopting pre-existing checkpoints "
                      f"with no identity stamp (legacy); stamping for next run.",
                      file=sys.stderr)
            return {"status": status, "archived_to": None, "warnings": warnings}

        # KEY match → adopt (refresh the stamp).
        if existing.get("key_digest") == fingerprint.get("key_digest"):
            self._write_fingerprint(fingerprint)
            return {"status": "match", "archived_to": None, "warnings": warnings}

        # KEY mismatch → backend identity changed. Archive and reset.
        archived_to = self._archive_dir(fingerprint.get("key_digest", ""))
        os.makedirs(self.dir, exist_ok=True)
        self._write_fingerprint(fingerprint)
        msg = ("backend identity changed since last run; prior checkpoints were "
               "NOT adopted (they would be a silent false-negative). Archived to "
               f"{archived_to}. Re-running this phase.")
        warnings.append(msg)
        if verbose:
            print(f"[{self.step_name}] {msg}", file=sys.stderr)
        return {"status": "reset", "archived_to": archived_to,
                "warnings": warnings}

    def _write_fingerprint(self, fingerprint: dict) -> None:
        self.ensure_dir()
        payload = dict(fingerprint)
        payload["written_at"] = (
            datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"))
        write_json(os.path.join(self.dir, FINGERPRINT_FILE), payload)

    def _archive_dir(self, key_digest: str) -> str:
        """Move the current checkpoint dir aside — never delete it."""
        short = key_digest.split(":")[-1][:8] or "reset"
        base = self.dir.rstrip("/")
        dest = f"{base}.superseded-{short}"
        n = 1
        while os.path.exists(dest):
            dest = f"{base}.superseded-{short}-{n}"
            n += 1
        shutil.move(self.dir, dest)
        return dest

    def write_summary(
        self,
        total_units: int,
        completed: int,
        errors: int,
        error_breakdown: dict,
        phase: str = "in_progress",
        usage: dict | None = None,
    ):
        """Write/overwrite _summary.json in checkpoint dir.

        Called from the main thread (as_completed loop) — no lock needed.

        Args:
            total_units: Total units in the step.
            completed: Number of successfully completed units.
            errors: Number of errored units.
            error_breakdown: Dict of error_type -> count.
            phase: ``"in_progress"`` or ``"done"``.
            usage: Optional dict with ``input_tokens``, ``output_tokens``,
                ``cost_usd`` accumulated so far for this step.
        """
        self.ensure_dir()
        filepath = os.path.join(self.dir, SUMMARY_FILE)
        data = {
            "step": self.step_name,
            "phase": phase,
            "timestamp": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            "total_units": total_units,
            "completed": completed,
            "errors": errors,
            "error_breakdown": error_breakdown,
        }
        if usage is not None:
            data["usage"] = usage
        write_json(filepath, data)
    @staticmethod
    def read_summary(checkpoint_dir: str) -> dict | None:
        """Read _summary.json from a checkpoint directory.

        Returns:
            Parsed dict or None if not found / unreadable.
        """
        filepath = os.path.join(checkpoint_dir, SUMMARY_FILE)
        if not os.path.isfile(filepath):
            return None
        try:
            return read_json(filepath)
        except (json.JSONDecodeError, OSError):
            return None

    def cleanup(self):
        """Remove the checkpoint directory (call on successful completion)."""
        if os.path.isdir(self.dir):
            shutil.rmtree(self.dir)
            print(f"[{self.step_name}] Cleaned up checkpoints", file=sys.stderr)

    _safe_filename = staticmethod(safe_filename)

    @staticmethod
    def status(checkpoint_dir: str) -> dict:
        """Return accurate checkpoint status by reading actual checkpoint files.

        This is the single source of truth for checkpoint counts. The Go CLI
        calls this via ``python -m vulnfounder checkpoint-status`` instead of
        doing its own file scanning.

        Returns:
            Dict with keys: step, checkpoint_dir, completed, errors,
            total_files, total_units, phase, error_breakdown.
        """
        # Derive step name from directory name (e.g. "enhance_checkpoints" → "enhance")
        dir_name = os.path.basename(checkpoint_dir.rstrip("/"))
        step = dir_name.replace("_checkpoints", "") if dir_name.endswith("_checkpoints") else dir_name

        result = {
            "step": step,
            "checkpoint_dir": checkpoint_dir,
            "completed": 0,
            "errors": 0,
            "total_files": 0,
            "total_units": 0,
            "phase": "unknown",
            "error_breakdown": {},
        }

        if not os.path.isdir(checkpoint_dir):
            return result

        # Read _summary.json for total_units and phase
        summary = StepCheckpoint.read_summary(checkpoint_dir)
        if summary:
            result["total_units"] = summary.get("total_units", 0)
            result["phase"] = summary.get("phase", "unknown")

        # Read all checkpoint files and classify each
        completed = 0
        errors = 0
        error_breakdown = {}

        for filename in os.listdir(checkpoint_dir):
            if not filename.endswith(".json") or filename in _RESERVED_FILES:
                continue
            filepath = os.path.join(checkpoint_dir, filename)
            try:
                data = read_json(filepath)
            except (json.JSONDecodeError, OSError):
                errors += 1
                error_breakdown["unreadable"] = error_breakdown.get("unreadable", 0) + 1
                continue

            unit_id = data.get("id")
            if not unit_id:
                errors += 1
                error_breakdown["missing_id"] = error_breakdown.get("missing_id", 0) + 1
                continue

            # Check for errors. Each phase stores checkpoint data differently:
            #   - enhance: error under agent_context (agentic) / llm_context (single-shot)
            #   - analyze: result.verdict == "ERROR" or result.finding == "error"
            #   - verify: verification is empty or verification.correct_finding == "error"
            #   - dynamic-test: top-level status == "ERROR"
            is_error = False
            err_type = None

            # Enhance-style: error under the context named by ``context_key``
            # (agent_context for agentic, llm_context for single-shot). Legacy
            # agentic checkpoints omit context_key → default agent_context.
            ctx_key = data.get("context_key", "agent_context")
            enhance_ctx = data.get(ctx_key, {})
            if isinstance(enhance_ctx, dict) and enhance_ctx.get("error"):
                is_error = True
                err = enhance_ctx["error"]
                err_type = err.get("type", "unknown") if isinstance(err, dict) else "unknown"

            # Analyze-style: result.verdict or result.finding
            elif "result" in data:
                res = data.get("result", {})
                if res.get("verdict") == "ERROR" or res.get("finding") == "error":
                    is_error = True
                    err_type = "analysis_error"

            # Verify-style: verification empty or correct_finding == "error"
            elif "verification" in data:
                v = data.get("verification", {})
                if not v or v.get("correct_finding") == "error":
                    is_error = True
                    err_type = "verification_error"

            # Dynamic-test-style: top-level status == "ERROR"
            elif data.get("status") == "ERROR":
                is_error = True
                err_type = "test_error"

            if is_error:
                errors += 1
                if err_type:
                    error_breakdown[err_type] = error_breakdown.get(err_type, 0) + 1
            else:
                completed += 1

        result["completed"] = completed
        result["errors"] = errors
        result["total_files"] = completed + errors
        result["error_breakdown"] = error_breakdown

        return result


def auto_checkpoint_dir(output_path: str, step_name: str) -> str:
    """Derive the checkpoint directory from the output file path.

    For enhance: output_path is dataset_enhanced.json
        -> same dir / enhance_checkpoints/
    For analyze: output_dir contains results.json
        -> output_dir / analyze_checkpoints/
    For verify: output_dir contains results_verified.json
        -> output_dir / verify_checkpoints/
    """
    if os.path.isdir(output_path):
        return os.path.join(output_path, f"{step_name}_checkpoints")
    return os.path.join(os.path.dirname(os.path.abspath(output_path)),
                        f"{step_name}_checkpoints")
